### ISSUE TITLE
Thoughts on input fields

### ISSUE THREAD

**Issue by aadcg** [atlas-engineer/nyxt#1020]
Hi Nyxt,

I'd like to work on the things I'll be addressing below but I would to
listen to you first.

Nyxt (for me) is the Emacs of web browsers.  That means that I must be
able to set my own keybindings when I'm editing text.

On commit 7c093434, John Mercouris introduced input-edit-mode.  Up to
that point, that functionality was bundled with web-mode.  The cursor
movements functions weren't set to any keys by default since for example
C-b (cursor-backwards) would conflict with history-backwards.

So this distinction makes sense but, in my opinion, we want
input-edit-mode getting toggled on/off depending if there's an active
input area.  So I think what I'm suggesting is a clear distinction
between reading a web page and editing it.  If you think of eww, the
keybindings to navigate pages are just letters of the alphabet because
there's this clear distinction.

This would mean that most of the commands that are currently binded to
control or meta keys could just be simplified.

What are your thoughts?

**Comment by Ambrevar** [comment]
Absolutely, so I think input-edit-mode can be automatically toggled
depending on the HTML element that's selected.

We already do this for vi-insert-mode, see:

```lisp
(define-command vi-button1 (&optional (buffer (current-buffer)))
  "Enable VI insert mode when focus is on an input element on the web page."
  ;; First we generate a button1 event so that the web view element is clicked
  ;; (e.g. a text field gets focus).
  (ffi-generate-input-event
   (current-window)
   (nyxt::last-event buffer))
  (let ((response (nyxt/web-mode:%clicked-in-input?)))
    (cond
      ((and (nyxt/web-mode:input-tag-p response)
            (find-submode buffer 'vi-normal-mode))
       (vi-insert-mode))
      ((and (not (nyxt/web-mode:input-tag-p response))
            (find-submode buffer 'vi-insert-mode))
       (vi-normal-mode)))))
```

@jmercouris Thoughts?

**Comment by jmercouris** [comment]
I thought about doing the same, but figured making a request before every command was expensive... I would like to make the checking more intelligent so that it pings less (and not for every command) or have it be event driven. The renderer tells us when we are in an input field, and when we are not.

**Comment by Ambrevar** [comment]
With VI we are not checking before every command, we are toggle the mode
on mouse-click, which is essentially the event-driven approach you are
talking about.

Misunderstanding?

**Comment by divansantana** [comment]
This would be nice, and make it possible to do smarter bindings similar to the default Emacs bindings with say EWW.

In EWW on google.com typing "g" in the input search field types a g, but typing g anyway else on the page, will call reload-buffer.

Currently the binding on nyxt is C-r.

**Comment by Ambrevar** [comment]
@jmercouris: What do you think of my proposal with regard to binding
"button1" like in VI mode?

**Comment by aadcg** [comment]
I need to dig a bit deeper but there are my thoughts. This "button1" functionality of `vi-mode` is of utmost importance for `emacs-mode`. So it seems that both modes would share this mechanism, and then they would differ in the keybindings.

It's interesting that @divansantana opened #1118 and expressed his comment above. It was exactly with that in mind that I opened this issue!

**Comment by Ambrevar** [comment]
André Alexandre Gomes <notifications@github.com> writes:

> I need to dig a bit deeper but these are my thoughts. This "button1" functionality of `vi-mode` is of utmost importance for `emacs-mode`. So it seems that both modes would share this mechanism, and then they would differ in the keybindings.

Agreed!

**Comment by aadcg** [comment]
Am I right that for those who use the `vi` keyscheme, `input-edit-mode` doesn't get triggered automatically on inputs fields?

**Comment by aartaka** [comment]
> Am I right that for those who use the `vi` keyscheme, `input-edit-mode` doesn't get triggered automatically on inputs fields?

Neither `input-edit-mode`, nor `vi-insert-mode` (which I'm guessing you meant instead of `input-edit-mode`) are not triggered on inputs, right. And WebKit does not provide us any signal to attach to for this behavior to work :(

Actually, we can do it in a slightly non-portable way: add a Parenscript to `on-signal-button-press-event` to check whether there's an active element in the page, and enabling `vi-insert-mode` if there is :thinking:

**Comment by aadcg** [comment]
It seems that on version 2.1.0, according to the `changelog`:

> vi mode automatically switches to vi-insert-mode when logical.

I agree with your suggestion.

**Comment by aartaka** [comment]
Note that right now it gets triggered on `button1` click, which causes problems like #2000...

**Comment by aadcg** [comment]
I see!

**Comment by brosasaki** [comment]
Bumping this to say there is still interest in this feature!

**Comment by aadcg** [comment]
Yes, this is really important. At the time I didn't have enough political weight to influence on a decision like this, but now I have. Patches welcome.
